ext/random/tests/03_randomizer/compatibility_user.phpt

```diff
@@ -1,80 +1,43 @@
-    {
-    }
+use Random\Engine;
+use Random\Engine\Mt19937;
+use Random\Engine\PcgOneseq128XslRr64;
+use Random\Engine\Test\TestWrapperEngine;
+use Random\Engine\Xoshiro256StarStar;
+use Random\Randomizer;
 
-    public function generate(): string
-    {
-        return $this->engine->generate();
-    }
-});
-for ($i = 0; $i < 1000; $i++) {
-    $native = $native_randomizer->nextInt();
-    $user = $user_randomizer->nextInt();
-    if ($native !== $user) {
-        die("failure Mt19937 i: {$i} native: {$native} user: {$user}");
-    }
-}
+include __DIR__ . "/../engines.inc";
 
-try {
-    $native_randomizer = new \Random\Randomizer(new \Random\Engine\PcgOneseq128XslRr64(1234));
-    $user_randomizer = new \Random\Randomizer(new class () implements \Random\Engine {
-        public function __construct(private $engine = new \Random\Engine\PcgOneseq128XslRr64(1234))
-        {
-        }
+$engines = [];
+$engines[] = new Mt19937(1234);
+$engines[] = new PcgOneseq128XslRr64(1234);
+$engines[] = new Xoshiro256StarStar(1234);
 
-        public function generate(): string
-        {
-            return $this->engine->generate();
-        }
-    });
-    
-    for ($i = 0; $i < 1000; $i++) {
-        $native = $native_randomizer->nextInt();
-        $user = $user_randomizer->nextInt();
-        if ($native !== $user) {
-            die("failure PcgOneseq128XslRr64 i: {$i} native: {$native} user: {$user}");
-        }
-    }
-} catch (\Random\RandomException $e) {
-    if ($e->getMessage() !== 'Generated value exceeds size of int') {
-        throw $e;
-    }
-}
+foreach ($engines as $engine) {
+    echo $engine::class, PHP_EOL;
 
-try {
-    $native_randomizer = new \Random\Randomizer(new \Random\Engine\Xoshiro256StarStar(1234));
-    $user_randomizer = new \Random\Randomizer(new class () implements \Random\Engine {
-        public function __construct(private $engine = new \Random\Engine\Xoshiro256StarStar(1234))
-        {
-        }
+    $native_randomizer = new Randomizer(clone $engine);
+    $user_randomizer = new Randomizer(new TestWrapperEngine(clone $engine));
+
+    for ($i = 0; $i < 10_000; $i++) {
+        $native = $native_randomizer->getInt(0, $i);
+        $user = $user_randomizer->getInt(0, $i);
 
-        public function generate(): string
-        {
-            return $this->engine->generate();
-        }
-    });
-    
-    for ($i = 0; $i < 1000; $i++) {
-        $native = $native_randomizer->nextInt();
-        $user = $user_randomizer->nextInt();
         if ($native !== $user) {
-            die("failure Xoshiro256StarStar i: {$i} native: {$native} user: {$user}");
+            die("failure: at {$i} - native: {$native} user: {$user}");
         }
     }
-} catch (\Random\RandomException $e) {
-    if ($e->getMessage() !== 'Generated value exceeds size of int') {
-        throw $e;
-    }
 }
 
 die('success');
+
 ?>
 --EXPECT--
+Random\Engine\Mt19937
```

Comment:
This actually fails for Mt19937, but this was not detected in the previous version of the test:

Mt19937 is a 32-bit engine (unless the requested range is larger than 32 bits), whereas userland engines are always 64-bit engines. Thus they use different `rand_rangeXX` internally, failing the test. The previous version did not detect this, because of it using `nextInt()` which is not affected.

Not sure what the best course of action is here.

Reply:
In a 32-bit environment, perhaps the user-defined engine should be used as 32-bit wide only for nextInt. Perhaps this will help maintain compatibility.

There may be further omissions to consider. I will mull it over.

Reply:
It was getting late yesterday. I'm already looking into this right now: It likely makes most sense to use `rand_range32` even with 64 bit engines if the requested range fits in 32 bits. That way the selected range implementation only depends on the user input parameters. This wastes some entropy of the upper 32 bits for the 64 bit engine, but we effectively require proper randomness in all bits already for getBytes() to work properly and Xoshiro/PCG don't suffer from low entropy in the lower bits anyway.

Reply:
Indeed, it seems to make sense. Apparently I implemented it that way initially, and I can't quite remember why I went with the current implementation...

https://github.com/php/php-src/pull/6568/files#diff-323e97ba9fdce2775b72efd5e1fedc52a3787e36c84e7bcd6b62bf48067cff7eR170

It seems likely that we followed the existing implementation without much deeper consideration. Thank you for all your accurate remarks.

Reply:
Fixed with #9418.